File: python/Bit Manipulation/sum_of_two_integers.py
class Solution:
    def getSum(self, a: int, b: int) -> int:
        """
        Approach:
        - Use bitwise operations to add two integers without using the '+' operator.
        - sum without carry = a ^ b
        - carry = (a & b) << 1
        - Repeat the process until there is no carry.
        - Use a mask to handle 32-bit integer overflow and negative numbers in Python.

        Time Complexity: O(1) (at most 32 iterations for 32-bit integers)
        Space Complexity: O(1)
        """
        mask = 0xffffffff
        while (mask & b) > 0:
            a, b = a ^ b, (a & b) << 1
        # Handle negative numbers
        return (mask & a) if a <= 0x7fffffff else ~(mask ^ a)

# -------------------
# Test Cases
# -------------------
def test_get_sum():
    sol = Solution()
    # Test case 1: Example from Leetcode
    assert sol.getSum(1, 2) == 3
    # Test case 2: Negative and positive
    assert sol.getSum(-2, 3) == 1
    # Test case 3: Both negative
    assert sol.getSum(-1, -1) == -2
    # Test case 4: Zero
    assert sol.getSum(0, 0) == 0
    # Test case 5: Large numbers
    assert sol.getSum(123456, 654321) == 777777
    print("All test cases passed!")

if __name__ == "__main__":
    test_get_sum()

File: C++/Bit Manipulation/sum_of_two_integers.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Approach:
     * - Use bitwise operations to compute sum without using '+'.
     * - At each step:
     *     sum without carry = a ^ b
     *     carry = (a & b) << 1
     * - Repeat until carry becomes 0.
     * - In C++, integers are already 32-bit and handle negative numbers with 2's complement,
     *   so no explicit mask is needed unlike Python.
     *
     * Time Complexity: O(1) → at most 32 iterations for a 32-bit integer.
     * Space Complexity: O(1)
     */
    int getSum(int a, int b) {
        while (b != 0) {
            unsigned carry = (unsigned)(a & b) << 1; // Calculate carry and shift left
            a = a ^ b;  // Sum without carry
            b = carry;  // New carry
        }
        return a;
    }
};

/*
📖 Line-by-line explanation:
----------------------------
#include <bits/stdc++.h>
    → Includes all standard C++ libraries.

using namespace std;
    → Avoids writing `std::` before every standard function/type.

class Solution { … };
    → Defines a class named `Solution` (LeetCode convention).

public:
    → Marks the following methods as accessible from outside the class.

int getSum(int a, int b) { … }
    → Function that takes two integers `a` and `b` and returns their sum without using `+`.

while (b != 0) {
    → Continue the loop until there is no carry left.

unsigned carry = (unsigned)(a & b) << 1;
    → Calculate the carry by ANDing a and b, then left-shifting it by 1.

a = a ^ b;
    → Compute the sum without carry using XOR.

b = carry;
    → Assign carry to b to handle it in the next iteration.

return a;
    → When carry becomes 0, `a` holds the final sum.

-----------------------------------------
⏱ Time Complexity: O(1) → Constant, at most 32 steps.
🗃 Space Complexity: O(1) → Constant space usage.
*/

